File: docs/validation/go/app/scenarios/s04_write_modify.go
package scenarios

// S-04: Write modified config and verify field change survives roundtrip
//
// ID:       S-04
// Action:   Read reference fixture, change machine_name, write to temp, re-read.
// Expected: machine_name change persists.

import (
	"fmt"
	"os"
	"path/filepath"

	mc "machine-config-go"
)

func RunS04WriteModify(fixturesDir, _ string) (bool, string) {
	path := filepath.Join(fixturesDir, "reference_config.h5")
	cfg, err := mc.NewReader(path).Parse()
	if err != nil {
		return false, fmt.Sprintf("read failed: %v", err)
	}

	cfg.Meta.MachineName = "MODIFIED_VALIDATION"
	cfg.Machine.MachineName = "MODIFIED_VALIDATION"

	tmpPath, err := tempH5("mcl_go_s04")
	if err != nil {
		return false, fmt.Sprintf("tempfile: %v", err)
	}
	defer os.Remove(tmpPath)

	if err := mc.NewWriter().Write(cfg, tmpPath); err != nil {
		return false, fmt.Sprintf("write failed: %v", err)
	}
	rb, err := mc.NewReader(tmpPath).Parse()
	if err != nil {
		return false, fmt.Sprintf("readback failed: %v", err)
	}
	if rb.Meta.MachineName != "MODIFIED_VALIDATION" {
		return false, fmt.Sprintf("meta.machine_name after roundtrip: %q", rb.Meta.MachineName)
	}
	if rb.Machine.MachineName != "MODIFIED_VALIDATION" {
		return false, fmt.Sprintf("machine.machine_name after roundtrip: %q", rb.Machine.MachineName)
	}
	return true, "machine_name modified and survives roundtrip"
}

File: docs/validation/rust/app/src/scenarios/s04_write_modify.rs
// S-04: Write modified config and verify field change survives round-trip
//
// ID:          S-04
// Action:      Read → change machine_name to "VALIDATION_TEST_MACHINE" →
//              write to temp file → read temp file → assert machine_name matches
// Expected:    machine_name persisted; file_version and build_plate_x unchanged.

use machine_config::{MachineConfigReader, MachineConfigWriter};
use std::path::Path;

pub fn run(fixtures_dir: &Path, _real_dir: &Path) -> (bool, String) {
    let path = fixtures_dir.join("reference_config.h5");
    let cfg = match MachineConfigReader::open(&path).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("read failed: {e}")),
    };
    let orig_x = cfg.machine.build_plate_x;

    let mut modified = cfg.clone();
    modified.meta.machine_name = "VALIDATION_TEST_MACHINE".to_string();
    modified.machine.machine_name = "VALIDATION_TEST_MACHINE".to_string();

    let tmp = match tempfile::Builder::new().suffix(".h5").tempfile() {
        Ok(t) => t,
        Err(e) => return (false, format!("tempfile creation failed: {e}")),
    };
    if let Err(e) = MachineConfigWriter::new(&modified).write(tmp.path()) {
        return (false, format!("write failed: {e}"));
    }

    let rb = match MachineConfigReader::open(tmp.path()).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("readback failed: {e}")),
    };

    if rb.meta.machine_name != "VALIDATION_TEST_MACHINE" {
        return (false, format!("machine_name not persisted: '{}'", rb.meta.machine_name));
    }
    if rb.meta.file_version.trim() != "1.0" {
        return (false, format!("file_version changed: '{}'", rb.meta.file_version));
    }
    if rb.machine.build_plate_x != orig_x {
        return (
            false,
            format!("build_plate_x changed: {orig_x:?} -> {:?}", rb.machine.build_plate_x),
        );
    }

    (true, "machine_name persisted, file_version and other fields unchanged".to_string())
}
